app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb

```diff
@@ -0,0 +1,515 @@
+      networks = collector.collect_networks
+      switchs(persister_host, host, nics, networks)
+      host_hardware(persister_host, host, networks, nics)
+    end
+  end
+
+  def host_hardware(persister_host, host, networks, nics)
+    hdw = host.cpu
+
+    memory_total_attr = host.statistics.to_miq_a.detect { |stat| stat.name == 'memory.total' }
+    memory_total = memory_total_attr && memory_total_attr.dig(:values, :first, :datum)
+    hw_info = host.hardware_information
+
+    cpu_cores = hdw.dig(:topology, :cores) || 1
+    cpu_sockets = hdw.dig(:topology, :sockets) || 1
+
+    persister_hardware = persister.hosts.find_or_build(persister_host).assign_attributes(
+      :cpu_speed            => hdw.speed,
+      :cpu_type             => hdw.name,
+      :memory_mb            => memory_total.nil? ? 0 : memory_total.to_i / 1.megabyte,
+      :cpu_cores_per_socket => cpu_cores,
+      :cpu_sockets          => cpu_sockets,
+      :cpu_total_cores      => cpu_sockets * cpu_cores,
+      :manufacturer         => hw_info.manufacturer,
+      :model                => hw_info.product_name
+    )
+
+    host_networks(persister_hardware, nics)
+    host_guest_devices(persister_hardware, nics, networks)
+  end
+
+  def host_networks(persister_hardware, nics)
+    nics.to_miq_a.each do |vnic|
+      # Get the ip section
+      ip = vnic.ip.presence || {}
+
+      persister.host_nics.find_or_build(persister_hardware).assign_attributes(
+        :description => vnic.name,
+        :ipaddress   => ip.address,
+        :subnet_mask => ip.netmask,
+      )
+    end
+  end
+
+  def host_guest_devices(persister_hardware, nics, networks)
+    nics.to_miq_a.each do |nic|
+      network = network_from_nic(nic, host, networks)
+
+      if network
+        switch_uid = network.id
+      else
+        switch_uid = network_name unless network_name.nil?
+      end
+
+      location = nil
+      location = $1 if nic.name =~ /(\d+)$/
+
+      persister.guest_devices.find_or_build(persister_hardware).assign_attributes(
+        :uid_ems         => nic.id,
+        :device_name     => nic.name,
+        :device_type     => 'ethernet',
+        :location        => location,
+        :present         => true,
+        :controller_type => 'ethernet',
+        :switch          => persister.switchs.lazy_find(switch_uid),
+        :network         => persister.host_nics.lazy_find(network.id)
+      )
+    end
+  end
+
+  def switchs(persister_host, host, nics, networks)
+    nics.to_miq_a.each do |nic|
+      network = network_from_nic(nic, host, networks)
+
+      persister_switch = persister.switchs.find_or_build(persister_host).assign_attributes(
+        :uid_ems => uid,
+        :name    => name,
+      )
+
+      lans(persister_switch, network)
+    end
+  end
+
+  def network_for_nic(nic, host, networks)
+    network_id = nic.dig(:network, :id)
+    if network_id
+      network = networks.detect { |n| n.id == network_id }
```

Comment:
if you cycle this through a lot of hosts&nics, might be better to avoid .detect which is O(n)

so somewhere above you could do indexed_networks = networks.index_by(:id)
and here just do indexed_networks[:network_id], which is O(1)

although I don't see the network_for_nic called anywhere?